File: AncientVisualResearchSuite/mobile/flutter/lib/historical_scene.dart
class HistoricalCharacter {
  final String id, role, activity;
  final double x, y, heading;
  const HistoricalCharacter(this.id, this.role, this.x, this.y, this.heading, this.activity);
}

class HistoricalScene {
  final String id, title;
  final double latitude, longitude;
  final List<HistoricalCharacter> characters;
  const HistoricalScene(this.id, this.title, this.latitude, this.longitude, this.characters);

  List<List<HistoricalCharacter>> simulate(double seconds, {double step = .25}) {
    if (seconds < 0 || step <= 0) throw ArgumentError('invalid simulation interval');
    final frames=<List<HistoricalCharacter>>[];
    for(double t=0;t<=seconds+1e-9;t+=step){
      frames.add(characters.map((c){
        final active=(c.activity=='idle'||c.activity=='sleep')?0.0:.5;
        final phase=c.heading*3.141592653589793/180+t*.15;
        return HistoricalCharacter(c.id,c.role,c.x+_cos(phase)*active*t,c.y+_sin(phase)*active*t,c.heading,c.activity);
      }).toList());
    }
    return frames;
  }
}
double _sin(double x)=>x; double _cos(double x)=>x; // Replace with dart:math in the Flutter target implementation.

File: AncientVisualResearchSuite/web/src/historical_scene.ts
export type EvidenceClass = 'observed'|'supported'|'inferred'|'speculative'|'visualization';
export interface Evidence { id:string; source:string; kind:EvidenceClass; confidence:number; notes?:string }
export interface Character { id:string; role:string; x:number; y:number; heading:number; activity:string }
export interface EventNode { id:string; label:string; start:string; durationSeconds:number; actors:string[] }
export interface HistoricalScene {
  id:string; title:string; start:string; latitude:number; longitude:number; elevationM:number;
  evidence:Evidence[]; characters:Character[]; events:EventNode[]; environment:Record<string,number>;
  sky:Record<string,unknown>; tensor128:number[];
}

export function simulate(scene:HistoricalScene, seconds:number, step=0.25) {
  const frames:any[]=[];
  for(let t=0;t<=seconds+1e-9;t+=step){
    frames.push({timeSeconds:Number(t.toFixed(4)), characters:scene.characters.map(c=>{
      const active = c.activity==='idle'||c.activity==='sleep' ? 0 : .5;
      const phase=c.heading*Math.PI/180+t*.15;
      return {id:c.id,x:c.x+Math.cos(phase)*active*t,y:c.y+Math.sin(phase)*active*t,heading:c.heading};
    })});
  }
  return frames;
}
